mongo/single_result.go

```diff
@@ -28,6 +29,22 @@ type SingleResult struct {
 	reg *bsoncodec.Registry
 }
 
+// NewSingleResultFromBytes creates a SingleResult with an underlying Cursor pre-loaded with the provided contents.
+func NewSingleResultFromBytes(contents []byte) *SingleResult {
```

Comment:
I opted to only add `NewSingleResultFromBytes` and _not_ `NewCursorFromBytes`. The latter seemed unnecessary to mock responses to read operations. Unlike `BatchCursor`, `Cursor` and `SingleResult` are both under the `mongo` package, so you can just as easily access `Cursor` internals from `single_result.go`. Let me know if you think users would benefit from `NewCursorFromBytes`.

Reply:
I'm not sure I follow the rationale for not needing a way to create a `Cursor`. Consider the example use case for testing a set of functions that need to call `FindOne` and `Find` by defining a mock type that satisfies that interface:
```go
type finder interface {
	Find(context.Context, interface{}, ...*options.FindOptions) (*mongo.Cursor, error)
	FindOne(context.Context, interface{}, ...*options.FindOneOptions) *mongo.SingleResult
}

type mockFinder struct {
	result []byte
}

func (mf *mockFinder) FindOne(context.Context, interface{}, ...*options.FindOneOptions) *mongo.SingleResult {
	return mongo.NewSingleResultFromBytes(mf.result)
}

func (mf *mockFinder) Find(context.Context, interface{}, ...*options.FindOptions) (*mongo.Cursor, error) {
	return &mongo.Cursor{Current: mf.result}, nil
}

func test() {
	b, err := bson.Marshal(bson.D{{"my", "value"}})
	if err != nil {
		panic(err)
	}
	mf := &mockFinder{result: b}

	doFindOne(mf)
	doFind(mf)
}
```

Running `FindOne` and decoding the `SingleResult` behaves as expected:
```go
func doFindOne(f finder) {
	var res bson.D
	err := f.FindOne(context.TODO(), bson.D{}).Decode(&res)
	if err != nil {
		panic(err)
	}
	fmt.Println(res)
}
```
Prints `[{my value}]`.

Running `Find` and iterating the cursor doesn't:
```go
func doFind(f finder) {
	cursor, err := f.Find(context.TODO(), bson.D{})
	if err != nil {
		panic(err)
	}

	for cursor.Next(context.TODO()) {
		var res bson.D
		err := cursor.Decode(&res)
		if err != nil {
			panic(err)
		}
		fmt.Println(res)
	}
}
```
Panics:
```
goroutine 1 [running]:
go.mongodb.org/mongo-driver/mongo.(*Cursor).next(0xc0000a2720, 0x163e910, 0xc0000ba008, 0x1483f00, 0xc0000c6330)
        ./mongo-go-driver/mongo/cursor.go:126 +0xef
go.mongodb.org/mongo-driver/mongo.(*Cursor).Next(...)
        ./mongo-go-driver/mongo/cursor.go:81
main.doFind(0x163a6a0, 0xc0001b0030)
        ./main.go:843 +0x17f
main.test()
        ./main.go:817 +0x145
main.main()
        ./main.go:794 +0x25
exit status 2
```

To successfully mock an API that returns a `Cursor`, we still need a way to create a `Cursor` that behaves like one returned by `Find` or `Aggregate`.

Reply:
Yep, we definitely need a way of mocking `Cursor` behavior. I've added a `NewMockCursor` constructor.